beginner_source/blitz/tensor_tutorial.py

```diff
@@ -1,195 +1,189 @@
+# ~~~~~~~~~~~~~~~~~
 #
-# You can use standard NumPy-like indexing with all bells and whistles!
+# Tensor attributes describe their shape data type and where they live.
 
-print(x[:, 1])
+tnsr = torch.rand(3,4)
 
-###############################################################
-# Resizing: If you want to resize/reshape tensor, you can use ``torch.view``:
-x = torch.randn(4, 4)
-y = x.view(16)
-z = x.view(-1, 8)  # the size -1 is inferred from other dimensions
-print(x.size(), y.size(), z.size())
+print(f"Shape of tensor: {tnsr.shape}")
+print(f"Datatype of tensor: {tnsr.dtype}")
+print(f"Device tensor lives on: {tnsr.device}")
 
-###############################################################
-# If you have a one element tensor, use ``.item()`` to get the value as a
-# Python number
-x = torch.randn(1)
-print(x)
-print(x.item())
 
-###############################################################
-# **Read later:**
-#
+######################################################################
+# --------------
 #
-#   100+ Tensor operations, including transposing, indexing, slicing,
-#   mathematical operations, linear algebra, random numbers, etc.,
-#   are described
-#   `here <https://pytorch.org/docs/torch>`_.
-#
-# NumPy Bridge
-# ------------
+
+
+######################################################################
+# Tensor Operations
+# ~~~~~~~~~~~~~~~~~
 #
-# Converting a Torch Tensor to a NumPy array and vice versa is a breeze.
+# Over 100 tensor operations, including transposing, indexing, slicing,
+# mathematical operations, linear algebra, random sampling, and more are
+# comprehensively described
+# `here <https://pytorch.org/docs/stable/torch.html>`__.
 #
-# The Torch Tensor and NumPy array will share their underlying memory
-# locations (if the Torch Tensor is on CPU), and changing one will change
-# the other.
+# Each of them can be run on the GPU (at typically higher speeds than on a
+# CPU). If you’re using Colab, allocate a GPU by going to Edit > Notebook
+# Settings.
 #
-# Converting a Torch Tensor to a NumPy Array
-# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-a = torch.ones(5)
-print(a)
+# We move our tensor to the GPU if available
+if torch.cuda.is_available():
+  tnsr = tnsr.to('cuda')
+
+
+######################################################################
+# Try out some of the operations from the list. If you're familiar with the NumPy API, you'll find the Tensor API a breeze to use.
+#
 
 ###############################################################
+# **Standard numpy-like indexing and slicing:**
+
+tnsr = torch.ones(4, 4)
+tnsr[:,1] = 0
+print(tnsr)
+
+######################################################################
+# **Both of these are joining ops, but they are subtly different.**
+
+t1 = torch.cat([tnsr, tnsr], dim=1)
+t2 = torch.stack([tnsr, tnsr], dim=1)
+print(t1)
+print()
+print(t2)
+
+######################################################################
+# **Multiply op - multiple syntaxes**
```

Comment:
Again here, it's important to first introduce one syntax, and then tell the user that there's another syntax.
It's also really important to separate out the code for each into each line. `tnsr * tnsr == tnsr.mul(tnsr)` tries to be too clever, while intimidating a new user who has their guard up while doing the first tutorial of a new package.

------

Multiplying Tensors

 this computes an element-wise product of two tensors
print(tensor * tensor)

this also computes an element-wise product of two tensors, and is an alternative syntax
print(tnsr.mul(tnsr))

this computes a matrix product, often also called matrix multiplication
tnsr @ tnsr.T

this can also be written as 
tnsr.matmul(tnsr.T)